src/algebra/algebra/restrict_scalars.lean

```diff
@@ -0,0 +1,145 @@
+-/
+import algebra.algebra.basic
+
+/-!
+
+# The `restrict_scalars` type alias
+
+See the documentation attached to the `restrict_scalars` definition for advice on how and when to
+use this type alias. As described there, it is often a better choice to use the `is_scalar_tower`
+typeclass instead.
+
+## Main definitions
+
+* `restrict_scalars R S M`: the `S`-module `M` viewed as an `R` module when `S` is an `R`-algebra.
+* `restrict_scalars.linear_equiv : restrict_scalars R S M ≃ₗ[S] M`: the equivalence as an
+  `S`-module between the restricted and origianl space.
+* `restrict_scalars.alg_equiv : restrict_scalars R S A ≃ₐ[S] A`: the equivalence as an `S`-algebra
+   between the restricted and original space.
+
+## See also
+
+There are many similarly-named definitions elsewhere which do not refer to this type alias. These
+refer to restricting the scalar type in a bundled type, such as from `A →ₗ[R] B` to `A →ₗ[S] B`:
+
+* `linear_map.restrict_scalars`
+* `linear_equiv.restrict_scalars`
+* `alg_hom.restrict_scalars`
+* `alg_equiv.restrict_scalars`
+* `submodule.restrict_scalars`
+* `subalgebra.restrict_scalars`
+-/
+
+variables (R S M A : Type*)
+
+/-- If we put an `R`-algebra structure on a semiring `S`, we get a natural equivalence from the
+category of `S`-modules to the category of representations of the algebra `S` (over `R`). The type
+synonym `restrict_scalars` is essentially this equivalence.
+
+Warning: use this type synonym judiciously! Consider an example where we want to construct an
+`R`-linear map from `M` to `S`, given:
+```lean
+variables (R S M : Type*)
+variables [comm_semiring R] [semiring S] [algebra R S] [add_comm_monoid M] [module S M]
+```
+With the assumptions above we can't directly state our map as we have no `module R M` structure, but
+`restrict_scalars` permits it to be written as:
+```lean
+-- an `R`-module structure on `M` is provided by `restrict_scalars` which is compatible
+example : restrict_scalars R S M →ₗ[R] S := sorry
+```
+However, it is usually better just to add this extra structure as an argument:
+```lean
+-- an `R`-module structure on `M` and proof of its compatibility is provided by the user
+example [module R M] [is_scalar_tower R S M] : M →ₗ[R] S := sorry
+```
+The advantage of the second approach is that it defers the duty of providing the missing typeclasses
+`[module R M] [is_scalar_tower R S M]`. If some concrete `M` naturally carries these (as is often
+the case) then we have avoided `restrict_scalars` entirely. If not, we can pass
+`restrict_scalars R S M` later on instead of `M`.
+
+Note that this means we almost always want to state definitions and lemmas in the language of
+`is_scalar_tower` rather than `restrict_scalars`.
+
+An example of when one might want to use `restrict_scalars` would be if one has a vector space
+over a field of characteristic zero and wishes to make use of the `ℚ`-algebra structure. -/
+@[nolint unused_arguments]
+def restrict_scalars (R S M : Type*) : Type* := M
```

Comment:
Any motivation for the rename `module.restrict_scalars` → `restrict_scalars`? Feels like this one of the less useful `restrict_scalars` defs, compared to e.g. `linear_map.restrict_scalars`, and with the rename it is more likely to cause name conflicts.

Reply:
I didn't rename it, it's [already called `restrict_scalars`](https://leanprover-community.github.io/mathlib_docs/algebra/algebra/basic.html#restrict_scalars).

Reply:
I see, this rename actually happened in #4733 and it took a bit too long for me to catch on :P

Reply:
I'd argue `module.restrict_scalars` would be a worse name anyway; it's not limited to restricting scalars on modules, presumably it would work for restricting scalars on distrib_mul_actions etc too.